core/arch/arm/sm/sm.c

```diff
@@ -1,21 +1,62 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
- * Copyright (c) 2016, Linaro Limited
+ * Copyright (c) 2016-2020, Linaro Limited
  * Copyright (c) 2014, STMicroelectronics International N.V.
  */
 #include <arm.h>
 #include <compiler.h>
+#include <config.h>
 #include <kernel/misc.h>
+#include <kernel/thread.h>
 #include <platform_config.h>
 #include <sm/optee_smc.h>
 #include <sm/sm.h>
 #include <sm/std_smc.h>
 #include <string.h>
 #include "sm_private.h"
 
+enum sm_handler_ret __weak sm_platform_handler(struct sm_ctx *ctx __unused)
+{
+	return SM_HANDLER_PENDING_SMC;
+}
+
+static void smc_arch_handler(struct thread_smc_args *args)
+{
+	uint32_t smc_fid = args->a0;
+	uint32_t feature_fid = args->a1;
+
+	switch (smc_fid) {
+	case ARM_SMCCC_VERSION:
+		args->a0 = SMCCC_V_1_1;
+		break;
+	case ARM_SMCCC_ARCH_FEATURES:
+		switch (feature_fid) {
+		case ARM_SMCCC_VERSION:
+		case ARM_SMCCC_ARCH_SOC_ID:
+			args->a0 = ARM_SMCCC_RET_SUCCESS;
+			break;
+		default:
+			args->a0 = ARM_SMCCC_RET_NOT_SUPPORTED;
+			break;
+		}
+		break;
+	case ARM_SMCCC_ARCH_SOC_ID:
+		args->a0 = ARM_SMCCC_RET_NOT_SUPPORTED;
+		break;
+	case ARM_SMCCC_ARCH_WORKAROUND_1:
+	case ARM_SMCCC_ARCH_WORKAROUND_2:
+		args->a0 = ARM_SMCCC_RET_NOT_REQUIRED;
+		break;
+	default:
+		args->a0 = OPTEE_SMC_RETURN_UNKNOWN_FUNCTION;
+		break;
+	}
+}
+
 uint32_t sm_from_nsec(struct sm_ctx *ctx)
 {
-	uint32_t *nsec_r0 = (uint32_t *)(&ctx->nsec.r0);
```

Comment:
How about keeping this variable, but without the casting and funny looking `()`, and use `nsec_r0` below instead of `args_base`?

Reply:
fair. will do.